From 57c8ea8b70639bc2e135485a28ab9c7319930e5f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Thu, 3 Apr 2025 16:04:10 +0200
Subject: [PATCH 1680/1695] drivers: stm32_rng: update RNG support for STM32MP2
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The STM32MP25, STM32MP23 and STM32MP21 have specific features that differ
from older MPUs.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 7fbeb1090740f2b3ea723747ed85b5cf904aa82c drivers: stm32_rng: update RNG support for STM32MP2]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: I29e04172dcd592831721d4fb4d62e97303b2a562
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/453898
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/stm32_rng.c | 35 +++++++++++++++++++++++------------
 1 file changed, 23 insertions(+), 12 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index 4c7727d57..3b0bf01c0 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -47,11 +47,8 @@
 #define RNG_CR_CONFIG2_SHIFT	U(13)
 #define RNG_CR_CLKDIV		GENMASK_32(19, 16)
 #define RNG_CR_CLKDIV_SHIFT	U(16)
-#define RNG_CR_CONFIG1		GENMASK_32(25, 20)
 #define RNG_CR_CONFIG1_SHIFT	U(20)
 #define RNG_CR_CONDRST		BIT(30)
-#define RNG_CR_ENTROPY_SRC_MASK	(RNG_CR_CONFIG3 | RNG_CR_NISTC | \
-				 RNG_CR_CONFIG2 | RNG_CR_CONFIG1)
 
 #define RNG_SR_DRDY		BIT(0)
 #define RNG_SR_CECS		BIT(1)
@@ -75,8 +72,7 @@
 #define RNG_FIFO_BYTE_DEPTH	U(16)
 #define RNG_CONF_LEN		U(3)
 
-#define RNG_CONFIG_MASK		(RNG_CR_ENTROPY_SRC_MASK | RNG_CR_CED | \
-				 RNG_CR_CLKDIV)
+#define RNG_CONFIG_MASK		(RNG_CR_CED | RNG_CR_CLKDIV)
 
 #define DT_RNG_MAX_NIST_CONFIG	U(3)
 
@@ -86,6 +82,7 @@ struct stm32_rng_driver_data {
 	uint32_t cr;
 	uint32_t nscr;
 	uint32_t htcr;
+	uint32_t cr_config1_mask;
 	bool has_power_optim;
 	bool has_cond_reset;
 };
@@ -111,6 +108,12 @@ struct stm32_rng_instance {
 /* Expect at most a single RNG instance */
 static struct stm32_rng_instance *stm32_rng;
 
+static uint32_t stm32_rng_get_entropy_mask(void)
+{
+	return (RNG_CR_CONFIG3 | RNG_CR_NISTC |
+		RNG_CR_CONFIG2 | stm32_rng->ddata->cr_config1_mask);
+}
+
 static vaddr_t get_base(void)
 {
 	assert(stm32_rng);
@@ -315,6 +318,7 @@ TEE_Result stm32_rng_init(void)
 	vaddr_t rng_base = get_base();
 	uint64_t timeout_ref = 0;
 	uint32_t cr_ced_mask = 0;
+	uint32_t entropy_mask = stm32_rng_get_entropy_mask();
 
 	res = enable_rng_clock();
 	if (res)
@@ -347,7 +351,7 @@ TEE_Result stm32_rng_init(void)
 		 */
 		if (!stm32_rng->rng_config)
 			stm32_rng->rng_config = io_read32(rng_base + RNG_CR) &
-						RNG_CR_ENTROPY_SRC_MASK;
+						entropy_mask;
 
 		/*
 		 * Configuration must be set in the same access that sets
@@ -355,7 +359,8 @@ TEE_Result stm32_rng_init(void)
 		 * not taken into account. CONFIGLOCK bit is always cleared at
 		 * this stage.
 		 */
-		io_clrsetbits32(rng_base + RNG_CR, RNG_CONFIG_MASK,
+		io_clrsetbits32(rng_base + RNG_CR,
+				RNG_CONFIG_MASK | entropy_mask,
 				stm32_rng->rng_config | RNG_CR_CONDRST |
 				cr_ced_mask |
 				(clock_div << RNG_CR_CLKDIV_SHIFT));
@@ -610,6 +615,7 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	struct dt_node_info dt_rng = { };
 	const fdt32_t *cuint = NULL;
 	int len = 0;
+	uint32_t entropy_mask = stm32_rng_get_entropy_mask();
 
 	fdt_fill_device_info(fdt, &dt_rng, node);
 	if (dt_rng.reg == DT_INFO_INVALID_REG)
@@ -647,8 +653,9 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	if (cuint && len > 0 &&
 	    (uint32_t)len <= DT_RNG_MAX_NIST_CONFIG * sizeof(uint32_t)) {
 		uint32_t i = 0;
+		uint32_t rng_cr_config1 = stm32_rng->ddata->cr_config1_mask;
 		uint32_t cr_shift_mask[DT_RNG_MAX_NIST_CONFIG][2] = {
-			{RNG_CR_CONFIG1_SHIFT, RNG_CR_CONFIG1},
+			{RNG_CR_CONFIG1_SHIFT, rng_cr_config1},
 			{RNG_CR_CONFIG2_SHIFT, RNG_CR_CONFIG2},
 			{RNG_CR_CONFIG3_SHIFT, RNG_CR_CONFIG3},
 		};
@@ -667,7 +674,8 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 
 	if (fdt_getprop(fdt, node, "st,rng-cfg-nist-custom", NULL))
 		stm32_rng->rng_config |= RNG_CR_NISTC;
-	if (stm32_rng->rng_config & ~RNG_CR_ENTROPY_SRC_MASK)
+
+	if (stm32_rng->rng_config & ~entropy_mask)
 		panic("Incorrect entropy source configuration");
 
 	cuint = fdt_getprop(fdt, node, "st,rng-htcfg", NULL);
@@ -745,6 +753,7 @@ static const struct stm32_rng_driver_data mp13_data[] = {
 		.nb_clock = 1,
 		.has_cond_reset = true,
 		.has_power_optim = true,
+		.cr_config1_mask = GENMASK_32(25, 20),
 		.cr = 0x00F00D00,
 		.nscr = 0x2B5BB,
 		.htcr = 0x969D,
@@ -767,6 +776,7 @@ static const struct stm32_rng_driver_data mp21_data[] = {
 		.nb_clock = 2,
 		.has_cond_reset = true,
 		.has_power_optim = true,
+		.cr_config1_mask = GENMASK_32(27, 20),
 		.cr = 0x00800D00,
 		.nscr = 0x01FF,
 		.htcr = 0xAAC7,
@@ -779,9 +789,10 @@ static const struct stm32_rng_driver_data mp25_data[] = {
 		.nb_clock = 2,
 		.has_cond_reset = true,
 		.has_power_optim = true,
-		.cr = 0x00F00D00,
-		.nscr = 0x2B5BB,
-		.htcr = 0x969D,
+		.cr_config1_mask = GENMASK_32(27, 20),
+		.cr = 0x08F01E00,
+		.nscr = 0x2E649,
+		.htcr = 0x6688,
 	},
 };
 
-- 
2.47.2

